File: src/FormingVectorModel/FormingVectorModel/Core/Objects/Visual/Image.cpp
#include "Image.h"

Core::Objects::Visual::Image::Image() : Image(0, 0)
{
}

Core::Objects::Visual::Image::Image(int rows, int cols) : _rows(rows), _cols(cols)
{
	_data = new Color[_rows * _cols];
}

Core::Objects::Visual::Image::~Image()
{
	if (_data)
		delete[] _data;
}

int Core::Objects::Visual::Image::rows()
{
	return _rows;
}

int Core::Objects::Visual::Image::cols()
{
	return _cols;
}

Core::Objects::Visual::IColor* Core::Objects::Visual::Image::get(int row, int col)
{
	if (row < _rows && col < _cols)
		return &_data[row * _cols + col];
	else
		throw "index out of range";
}

void Core::Objects::Visual::Image::set(int row, int col, IColor* value)
{
	if (row < _rows && col < _cols)
		_data[row * _cols + col] = Color(value->red(), value->green(), value->blue());
	else
		throw "index out of range";
}

void Core::Objects::Visual::Image::transform(Transformations::ITransformation* transforamtion)
{
	transforamtion->execute(this);
}

std::vector<Core::Objects::Geometric::IObject*> Core::Objects::Visual::Image::vectorize(Vectorizations::IVectorization* vectorization)
{
	return vectorization->execute(this);
}

std::vector<Core::Objects::Visual::IColor*> Core::Objects::Visual::Image::get_neighbours(int row, int col, unsigned char brightness)
{
	std::vector<IColor*> neighbours;

	for (int i = -1; i <= 1; i++)
	{
		for (int j = -1; j <= 1; j++)
		{
			if (i == 0 && j == 0) continue;

			int index = (row + i) * _cols + (col + j);
			if (0 <= row + i && row + i < _rows && 0 <= col + j && col + j < _cols &&
				_data[index].brightness() == brightness)
			{
				neighbours.push_back(&_data[index]);
			}
		}
	}

	return neighbours;
}

File: src/FormingVectorModel/FormingVectorModelLibrary/Core/Objects/Visual/Image.cs
using System.Collections;
using System.Collections.Generic;
using FormingVectorModelLibrary.Core.Objects.Mathematic;
using FormingVectorModelLibrary.Core.Transformation;

namespace FormingVectorModelLibrary.Core.Objects.Visual
{
    public class Image : IImage
    {
        private IMatrix<IColor> data;

        public Image(int rows, int cols)
        {
            data = new Matrix<IColor>(rows, cols);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="rows"></param>
        /// <param name="cols"></param>
        /// <param name="data">Array with BGR format color</param>
        public Image(int rows, int cols, byte[,,] data) : this(rows, cols)
        {
            for (int i = 0; i < Rows; i++)
            {
                for (int j = 0; j < Cols; j++)
                {
                    this.data[i, j] = new Color(data[i, j, 2], data[i, j, 1], data[i, j, 0]);
                }
            }
        }

        public Image() : this(0, 0) { }

        public int Rows => data.Rows;

        public int Cols => data.Cols;

        public IColor this[int row, int col]
        {
            get { return data[row, col]; }
            set { data[row, col] = value; }
        }

        /// <summary>
        /// Return 3d array. 1 - rows, 2 - cols, 3 - BGR values
        /// </summary>
        public byte[,,] Data
        {
            get
            {
                byte[,,] rowData = new byte[Rows, Cols, 3];

                for (int i = 0; i < Rows; i++)
                {
                    for (int j = 0; j < Cols; j++)
                    {
                        rowData[i, j, 0] = data[i, j].Blue;
                        rowData[i, j, 1] = data[i, j].Green;
                        rowData[i, j, 2] = data[i, j].Red;
                    }
                }

                return rowData;
            }
        }

        public void Transform(ITransformation transformation)
        {
            transformation.Execute(this);
        }

        public IEnumerator GetEnumerator()
        {
            return data.GetEnumerator();
        }

        public IList<IColor> GetNeighboursByBrightness(int row, int col, byte brightness)
        {
            IList<IColor> neighbours = new List<IColor>();
            for (int i = -1; i <= 1; i++)
            {
                for (int j = 0; j <= 1; j++)
                {
                    if (i == 0 && j == 0) continue;

                    
                }
            }

            return null;
        }
    }
}